### ISSUE TITLE
Exposing different globals via std::unique_ptr<S, py::nodelete> > can lead to pybind11_object_dealloc(): Tried to deallocate unregistered instance!

### ISSUE THREAD

**Issue by heiner** [pybind/pybind11#2394]
First of all, many thanks for this excellent library! <3

## Issue description

I want to expose the values of global structs defined in a library I'm using. As I didn't get `py::return_value_policy::reference` to work with `py::init` or `def("__init__", ...)`, I'm trying with a holder type of `std::unique_ptr<S, py::nodelete> >`. This appears to work at first, and in many situations, but breaks in a situation where different objects of this type override each other.

## Reproducible example code

```cpp
#include <pybind11/pybind11.h>

namespace py = pybind11;

struct S {
  const char name[80];
};

S sarray[2] = {
    {"hello"},
    {"world"},
};

PYBIND11_MODULE(skel, m) {
  py::class_<S, std::unique_ptr<S, py::nodelete> >(m, "S")
      .def(py::init([](int index) -> S* { return &sarray[index]; }))
      .def_readonly("name", &S::name);
}
```

Together with
```python
from skel import skel


def test_skel():
    for _ in range(2):  # range(1) works.
        s = skel.S(0)
        name0 = s.name


if __name__ == "__main__":
    test_skel()
```

On Ubuntu 18.04 with GCC 7.4.0, this fails with:

```
$ python main.py
terminate called after throwing an instance of 'std::runtime_error'
  what():  pybind11_object_dealloc(): Tried to deallocate unregistered instance!
Aborted
```

Interestingly, it works on MacOS 10.15.5 with Apple clang version 11.0.0 (clang-1100.0.33.12).

**Comment by heiner** [comment]
Update: The following code appears to work, but I'm not certain I'm holding this right.

```cpp
  py::class_<S>(m, "S")
      .def(
          "__init__",
          [](py::detail::value_and_holder& v_h, int index) {
            v_h.value_ptr() = &sarray[index];
            v_h.inst->owned = false;
            v_h.set_holder_constructed(true);
          },
          py::detail::is_new_style_constructor())
      .def_readonly("i", &S::i)
      .def_readonly("name", &S::name);
```

**Comment by bstaletic** [comment]
> Interestingly, it works on MacOS 10.15.5 with Apple clang version 11.0.0 (clang-1100.0.33.12).

It also works for me, on linux with gcc 10 or clang 10.

**Comment by bstaletic** [comment]
Actually, scratch that. It just happens at exit. I managed to reproduce eventually.

**Comment by YannickJadoul** [comment]
Is this related to #2252, #1592, and #1568? Let me try if that proposed fix helps.

**Comment by YannickJadoul** [comment]
See #2252 for further updates, @heiner & @bstaletic :-)

**Comment by bstaletic** [comment]
Fixed in 2.6.0. Thanks to @YannickJadoul and @henryiii
